src/Workspaces/Remote/Core/Diagnostics/PerformanceQueue.cs

```diff
@@ -0,0 +1,226 @@
+
+                Reset(_performanceMap, rawData, unitCount);
+            }
+
+            public void Update(IEnumerable<(string, TimeSpan)> rawData, int unitCount)
+            {
+                Reset(_performanceMap, Convert(rawData), unitCount);
+            }
+
+            public void AppendAnalyzers(HashSet<int> analyzerSet)
+            {
+                analyzerSet.UnionWith(_performanceMap.Keys);
+            }
+
+            public double? GetTimeSpanInMillisecond(int assignedAnalyzerNumber)
+            {
+                if (!_performanceMap.TryGetValue(assignedAnalyzerNumber, out var value))
+                {
+                    return null;
+                }
+
+                return value;
+            }
+
+            private void Reset(
+                Dictionary<int, double> map, IEnumerable<(int assignedAnalyzerNumber, TimeSpan timeSpan)> rawData, int fileCount)
+            {
+                // get smallest timespan in the snapshot
+                var minSpan = rawData.Select(kv => kv.timeSpan).Min();
+
+                // for now, we just clear the map, if reusing dictionary blindly became an issue due to
+                // dictionary grew too big, then we need to do a bit more work to determine such case
+                // and re-create new dictionary
+                map.Clear();
+
+                // map is normalized to current timespan - min timspan of the snapshot
+                foreach (var (assignedAnalyzerNumber, timeSpan) in rawData)
+                {
+                    map[assignedAnalyzerNumber] = (timeSpan.TotalMilliseconds - minSpan.TotalMilliseconds) / fileCount;
+                }
+            }
+
+            private static IEnumerable<(int, TimeSpan)> Convert(IEnumerable<(string analyzerId, TimeSpan timeSpan)> rawData)
+            {
+                return rawData.Select(kv => (AnalyzerNumberAssigner.Instance.GetUniqueNumber(kv.analyzerId), kv.timeSpan));
+            }
+        }
+
+        /// <summary>
+        /// Assign unique number to diagnostic analyzers
+        /// </summary>
+        private class AnalyzerNumberAssigner
+        {
+            public static readonly AnalyzerNumberAssigner Instance = new AnalyzerNumberAssigner();
+
+            private int _currentId;
+
+            // use simple approach for now. we don't expect it to grow too much. so entry added
+            // won't be removed until process goes away
+            private readonly Dictionary<string, int> _idMap;
+
+            private AnalyzerNumberAssigner()
+            {
+                _currentId = 0;
+                _idMap = new Dictionary<string, int>();
+            }
+
+            public int GetUniqueNumber(DiagnosticAnalyzer analyzer)
+            {
+                return GetUniqueNumber(analyzer.GetAnalyzerId());
+            }
+
+            public int GetUniqueNumber(string analyzerName)
+            {
+                if (!_idMap.TryGetValue(analyzerName, out var id))
```

Comment:
⚠️ This is a non-thread-safe access to an effectively static field. Make sure the behavior is what's intended.

Reply:
the outer type comments states that thread-safety should be handled by caller. not by this type.